File: src/Triggers/ShieldTrigger.cs
using Celeste.Mod.Entities;
using Celeste.Mod.GameHelper.Utils;
using Microsoft.Xna.Framework;

namespace Celeste.Mod.GameHelper.Triggers;

[CustomEntity("GameHelper/ShieldTrigger")]
public class ShieldTrigger : Trigger {
    private readonly bool enable;
    private readonly int flashes;

    public ShieldTrigger(EntityData data, Vector2 levelOffset) : base(data, levelOffset) {
        enable = data.Bool("enable");
        flashes = data.Int("flashAmount");
    }

    public override void OnEnter(Player p) {
        p.Scene.Entities.FindAll<Shield>().ForEach(s => s.RemoveSelf());
        if (enable) {
            SceneAs<Level>().Add(new Shield(flashes));
        }
    }
}

File: Loenn/triggers/private/shield_trigger.lua
local shield = {}

shield.name = "GameHelper/ShieldTrigger"
shield.depth = 8998
shield.justification = { 0.0, 0.0 }
shield.placements = {
    name = "shield",
    data = {
        flashAmount = 3,
        enable = true
    }
}
shield.fieldInformation = {
    flashAmount = {
        fieldType = "integer"
    }
}

return shield
